pkg/engine/deployment.go

```diff
@@ -182,8 +182,13 @@ func newDeployment(ctx *Context, info *deploymentContext, opts deploymentOptions
 					"Type tokens must be of the format <package>:<module>:<type> - "+
 					"refer to the import section of the provider resource documentation.", imp.Type.String())
 			}
-			if imp.Provider == "" && imp.Version == nil {
-				imp.Version = defaultProviderVersions[imp.Type.Package()]
+			if imp.Provider == "" {
+				if imp.Version == nil {
+					imp.Version = defaultProviderVersions[imp.Type.Package()].Version
+				}
+				if imp.ServerURL == "" {
```

Comment:
We'd discussed offline whether the name of the input property to the provider should be `serverURL` or `pluginDownloadURL` and decided to go with the latter. The latter is more verbose, but it aligns with the [name of the property in the schema](https://github.com/pulumi/pulumi/blob/c95e49e5abc9ea845f08475a3a44bcf9ea593942/pkg/codegen/schema/schema.go#L1520) and is less likely to conflict with a provider's own input properties (which very well could have a `serverURL` for its own purposes). This PR does use `pluginDownloadURL` for the input property itself, but everywhere else in the implementation is using `ServerURL`.

Two questions:

1. Should we use `PluginDownloadURL` throughout instead of `ServerURL` to be consistent and keep the concept count lower for easier readability/understandability of the code?
   - I realize we do use `ServerURL` in the [`workspace.PluginInfo`](https://github.com/pulumi/pulumi/blob/c95e49e5abc9ea845f08475a3a44bcf9ea593942/sdk/go/common/workspace/plugins.go#L93) struct and `Server` for the [`plugin.PulumiPluginJSON`](https://github.com/pulumi/pulumi/blob/eb735c7106fcbb48eb3e0edaab3d0baea47d2f0e/sdk/go/common/resource/plugin/plugin.go#L58) struct, so there is a case to be made for `ServerURL`.
1. When we expose this in the SDKs on `ResourceOptions`, do we want to expose it as `ResourceOptions.PluginDownloadURL` or `ResourceOptions.ServerURL`?

Curious if anyone else has a preference.

Reply:
1. @justinvp You convinced me that `"serverURL"` had a higher chance of conflicting with user inputs/outputs. It should be impossible to conflict with the `ServerURL` name in go, unless I'm confused. Given that we have (and are locked into through serialization) two different names for the same concept, I figured that I should choose a single name (for in the code base) and stick with it. I chose `serverURL` because 2/3 of our endpoints have that name. I would be happy to change it to `PulumiPluginJSON` if you would prefer. 

2. I think `ResourceOptions.PluginDownloadURL` is probably clearer there.

I would also be curious what others think.